src/tasks.ts

```diff
@@ -0,0 +1,359 @@
+        return;
+    }
+
+    const targetEpoch = await nextReportEpoch(blockTag);
+    debug('Target epoch is', targetEpoch);
+
+    if (targetEpoch > slotToEpoch(clBlock.message.slot)) {
+        debug('Target epoch is not reached yet');
+        return;
+    }
+
+    const [from, to] = await reportFrame(blockTag);
+    debug('Report frame in epochs is', [from, to]);
+
+    const artifact = readArtifact(from, to);
+    if (!R.isNil(artifact)) {
+        debug('Report tree artifact already exists');
+        return;
+    }
+
+    const prevTreeCID = await Shared.ORACLE.treeCid({
+        blockTag,
+    });
+
+    if (R.isEmpty(prevTreeCID)) {
+        debug('No previous report tree CID found');
+    } else {
+        debug('Reading the previous tree from IPFS by CID', prevTreeCID);
+        const prevTree = loadTree(await catCID(prevTreeCID));
+
+        debug('Check the tree root');
+        const prevRoot = await Shared.ORACLE.reportRoot({
+            blockTag,
+        });
+
+        if (prevTree.root !== prevRoot) {
+            throw new Error('Tree root mismatch');
+        }
+
+        debug('Fill in the rewards cache');
+        for (const [_, [no, reward]] of prevTree.entries()) {
+            Cache.rewards.set(no, reward);
+        }
+    }
+
+    debug('Loading validators');
+    await loadValidators(Number(targetEpoch));
+
+    debug('Fetching duties');
+    for (const epoch of iterEpochs(from, to)) {
+        await lookupCommittees(epoch);
+    }
+
+    debug('Check duties');
+    for (const epoch of iterEpochs(from, to)) {
+        await checkEpochSlots(epoch);
+    }
+
+    // TODO: read the threshold from somewhere
+    Cache.excludeFromStats(belowThreshold(0.9));
+
+    const feeShares = await Shared.STETH.sharesOf(Shared.CONFIG.CSM_ADDRESS);
+    const distributed = distributeFees(feeShares);
+
+    const leafs = R.sortBy(R.prop(0), [...Cache.rewards.entries()]);
+    debug('Rewards distribution', leafs);
+    const tree = buildTree(leafs);
+    debug('Report tree root', tree.root);
+
+    debug('Simulating the report');
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport.staticCall(
+        targetEpoch,
+        tree.root,
+        distributed,
+        '', // CID TBA, but can be precalculated
+    );
+    debug('Simulation successfull!');
+
+    debug('Uploading the tree to IPFS');
+    const cid = await uploadTree(tree);
+    debug('Tree uploaded, CID is', cid.toString());
+
+    debug('Sending the report');
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport(
+        targetEpoch,
+        tree.root,
+        distributed,
+        cid.toString(),
+    );
+
+    debug('Report sent, storing the artifact');
+    saveArtifact(from, to, tree);
+}
+
+async function nextReportEpoch(blockTag: BlockTag) {
+    const r = await Shared.ORACLE.nextReportEpoch({
+        blockTag,
+    });
+
+    if (!isUint64(r)) {
+        throw new Error('Next report epoch is too big');
+    }
```

Comment:
Well, there are typings from lodestar with `Epoch` and another types aliased to `Number`. That's why I decided to keep the separation, but it can be reconsidered.